packages/decoupled/src/config/config-validator.ts

```diff
@@ -0,0 +1,157 @@
+        },
+    ) {
+        this.validator = new Ajv(validatorOptions);
+
+        addKeywords(this.validator, ['typeof']);
+
+        this.contributes = this.loadContributes();
+        this.defsSchema = this.generateDefinitionsSchema();
+        this.schema = this.generateSchemaFromDefinitions();
+    }
+
+    get autoloadPattern() {
+        return `'**/${this.contributeFileName}'`;
+    }
+
+    /**
+     * Validate input data
+     * @param data
+     */
+    public validate(data: AnyObject): [ null | ErrorObject[] , boolean] {
+        const { defsSchema, schema } = this;
+
+        const validate = this.validator
+            .addSchema(defsSchema)
+            .compile(schema);
+        const valid = validate(data);
+
+        return [validate.errors, valid];
+    }
+
+    /**
+     * Generate definitions schema from decoupled contributes
+     */
+    private generateDefinitionsSchema(): DefsSchemaInterface {
+        const { contributes, TJSSettings, TJSCompilerOptions } = this;
+
+        const program = TJS.getProgramFromFiles(contributes, TJSCompilerOptions);
+
+        return TJS.generateSchema(program, '*', TJSSettings, contributes);
+    }
+
+    /**
+     * Generate actual schema from definitions schema
+     */
+    private generateSchemaFromDefinitions(): JSONSchemaInterface {
+        const { defsSchema: { definitions = {} } } = this;
+
+        const properties = Object.entries(definitions).reduce((init, [k, v]) => {
+            init[k.toLowerCase()] = { $ref: `decoupled.contrib.json#/definitions/${k}` };
+            return init;
+        }, {});
+
+        const required = Object.keys(properties)
+            .filter((k) => k[0] !== '$'); // Filter out common types with leading $
+
+        return {
+            type: 'object',
+            properties,
+            required,
+        };
+    }
+
+    /**
+     * Load contributes file locations
+     */
+    private loadContributes() {
+        const modules = getFromDecoupledConfig('modules');
+
+        if (modules && Array.isArray(modules) && modules.length > 0) {
+            return this.loadContributesFromModules(modules);
+        }
+
+        logger.warn('No modules config found in Decoupled Config.');
+
+        return this.autoLoadContributes();
+    }
+
+    /**
+     * Load contributes file locations from decoupled.config
+     *
+     * 1. Root and `src` directories
+     * 2. Sites directories
+     * 3. Modules directories under `node_modules`
+     */
+    private loadContributesFromModules(modules: string[]) {
+        const { contributeFileName } = this;
+
+        const srcDir = getFromDecoupledConfig('srcDir');
+        const projectContributes = [ './', srcDir ];
+
+        const siteIDs = getSiteIDs(resolve(srcDir, 'sites'), true);
+        const siteContributes = siteIDs.map((site) => join(srcDir, 'sites', site));
+
+        const modulesContributes = ['decoupled', ...modules].map((path) => join('node_modules', path));
```

Comment:
please don't test modules by assuming the location of node_modules. this might break i.e. for mono-repos or nested dependency trees. Use [require.resolve](https://nodejs.org/api/modules.html#modules_require_resolve_request_options) instead